File: java/34-FindFirstAndLastPositionOfElementInSortedArray.java
// 2022-04-22 21:35:29
// Runtime 1ms(10.6%)
// Memory 44.2mb(84.4%)

/**
 * 你连二分查找都不会，我为什么要给你机会
 */

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int i;
        int[] ret = {-1, -1};
        boolean find = false;
        for (i = 0; i < nums.length; i++) {
            if (nums[i] == target) {
                ret[0] = i;
                find = true;
                break;
            }
        }
        if (!find)
            return ret;
        for(i = i; i < nums.length; i++) {
            if (i == nums.length - 1 || nums[i + 1] != target) {
                ret[1] = i;
                break;
            }
        }
        return ret;
    }

File: cpp/34-FindFirstAndLastPositionOfElementInSortedArray.cpp
// 2022-05-15 17:55:04
// Runtime 8ms(58.3%)
// Memory 13.9mb(5.2%)

/**
 * 二分查找是吧 现在会了
 */

class Solution {
public:
    int binarySearch(vector<int> arr, int target, bool upper) {
        int l(-1), r(arr.size()), m;
        if (upper) {
            while (l + 1 != r) {
                m = (l + r) / 2;
                if (arr[m] <= target)
                    l = m;
                else
                    r = m;
            }
            return l;
        }
        else {
            while (l + 1 != r) {
                m = (l + r) / 2;
                if (arr[m] < target)
                    l = m;
                else
                    r = m;
            }
            return r;
        }
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        if (!nums.size())
            return {-1, -1};
        
        int upper, lower;
        upper = binarySearch(nums, target, true);
        lower = binarySearch(nums, target, false);
        if (upper < 0)
            upper = 0;
        
        if (nums[upper] != target ||
            nums[lower] != target)
            return {-1, -1};
        else
            return {lower, upper};
    }
};
